File: python/0048.py
class Solution(object):
    def rotate(self, matrix):
        """
        :type matrix: List[List[int]]
        :rtype: void Do not return anything, modify matrix in-place instead.
        """
        n = len(matrix)
        for i in range(n - 1):
            for j in range(i, n - 1 - i):
                t = matrix[i][j]
                x, y = i, j
                for k in range(3):
                    tx = n - y - 1
                    ty = x
                    matrix[x][y] = matrix[tx][ty]
                    x, y = tx, ty
                matrix[x][y] = t

File: rust/src/bin/0048.rs
struct Solution;


impl Solution {
    pub fn rotate(matrix: &mut Vec<Vec<i32>>) {
        if matrix.len() == 0 {
            return;
        }
        let sz = matrix.len();
        let get = |x: usize, y: usize| {
            (y, sz - 1 - x)
        };
        for x in 0..=sz / 2 {
            for y in x..sz - 1 - x {
                let mut temp = matrix[x][y];
                let (mut xx, mut yy) = (x, y);
                for i in 0..3 {
                    let (mut xxx, mut yyy) = get(xx, yy);
                    let nex = matrix[xxx][yyy];
                    matrix[xxx][yyy] = temp;
                    temp = nex;
                    xx = xxx;
                    yy = yyy;
                }
                matrix[x][y] = temp;
            }
        }
    }
}

fn main() {
    let mut a = vec![
        vec![1, 2, ],
        vec![4, 5, ],
    ];
    Solution::rotate(&mut a);
    println!("{:?}", a);
}